### PR TITLE
Práctica 5: Tech evolution. Caso POS Tagging

### PR DESCRIPTION
**Pull request by fwgalde**
Este PR incorpora la solución completa de la práctica 5; es decir:

- Obtención de los embeddings (_estáticos_) de 100 palabras al azar (con un _**muestreo estratificado**_, más info en el notebook) del modelo RNN visto en clase.
- Aplicación de un algoritmo de clusterización (Spectral Clustering, se probaron vairos) a las palabras y plotearlas en 2D (varias reducciones de dimensión para gráficar, como t-SNE, PCA y UMAP).
- Agrega al plot los embeddings de las etiquetas POS. A través de un promedio del submuestreo (se explica más en el notebook). 
- [_**Extra**_] Implementación de una red Long short-term memory units (LSTM) para la tarea de etiquetado POS, su reporte de accuracy y comparación con los resultados de los demás métodos.

A veces siento que me tardo mucho, una disculpa. Espero por lo menos estén bien. Felices vacaciones 🌴🌴

### CODE REVIEW

File: practicas/fwgalde/P5/5_tech_evolution.py [hunk 1]
```diff
@@ -0,0 +1,2170 @@
...
+
+"""# Práctica 5: Tech evolution. Caso POS Tagging
+
+**Fecha de entrega: 13 de Abril 2025 11:59pm**
+
+## ***Ejercicio 1***
+- Obten los embeddings de 100 palabras al azar del modelo RNN visto en clase
+  - Pueden ser los embeddings estáticos o los dinámicos del modelo
+
+### ***Stratified random sampling of embeddings***
```
Comment (umoqnier) [inline practicas/fwgalde/P5/5_tech_evolution.py:1002]: Interesante como abordaste la solución de esto

File: practicas/fwgalde/P5/5_tech_evolution.py [hunk 2]
```diff
@@ -0,0 +1,2170 @@
...
+        elif model_name == 'SpectralClustering':
+            model = SpectralClustering(n_clusters=k, random_state=RANDOM_STATE, affinity="nearest_neighbors")
+        elif model_name == 'HDBSCAN':
+            model = hdbscan.HDBSCAN(min_cluster_size=k)
+        elif model_name == 'OPTICS':
+            model = OPTICS(min_samples=k)
+        elif model_name == 'DBSCAN':
+            model = DBSCAN(min_samples=k)
+        else:
+            model = KMeans(n_clusters=k, random_state=RANDOM_STATE, n_init='auto')
```
Comment (umoqnier) [inline practicas/fwgalde/P5/5_tech_evolution.py:1098]: Quizá probaste demasiados o más bien entregaste demasiadas opciones. Considero que está bueno la exploración, pero a la entrega hubieras elegido un par y justificado porque para no atascar tanto tu entrega. Igual está interesante ver los resultados

File: practicas/fwgalde/P5/5_tech_evolution.py [hunk 3]
```diff
@@ -0,0 +1,2170 @@
...
+dbscan = DBSCAN(min_samples=2)
+dbscan_labels = dbscan.fit_predict(word_embeds)
+
+dbscan_counts = Counter(dbscan_labels)
+rprint(dbscan_counts)
+
+metrics(dbscan_labels, word_embeds)
+plot_clustering(word_embeds, dbscan_labels, "DBSCAN")
+
+"""El modelo Spectral Clustering con $k=2$ emerge como la opción óptima para la tarea de clustering basado en embeddings y etiquetas POS. Este método presenta un equilibrio destacado entre métricas de evaluación, con un Silhouette Score de 0.207 (mayor cohesión intra-cluster), Calinski-Harabasz de 20.0 (mejor separación global) y Davies-Bouldin de 1.92 (balance entre compactación y distancia inter-cluster).
```
Comment (umoqnier) [inline practicas/fwgalde/P5/5_tech_evolution.py:1372]: Genial, veo que justo fuiste exhaustivo en este apartado. Me agrada

File: practicas/fwgalde/P5/5_tech_evolution.py [hunk 4]
```diff
@@ -0,0 +1,2170 @@
...
+
+plt.tight_layout()
+plt.show()
+
+"""## ***Ejercicio 4***
+- Realiza una conclusión sobre los resultados observados
+
+### ***Conclusión***
+
+La proyección no lineal de embeddings mediante t-SNE y UMAP valida numérica y visualmente la efectividad de Spectral Clustering (*k* = 2) para identificar una partición robusta en el espacio sintáctico-semántico. Mientras PCA, al priorizar varianza global, no logra separar claramente las clases, t-SNE y UMAP revelan dos clusters cohesivos centrados en los centroides POS, confirmando que la estructura subyacente en los embeddings refleja patrones lingüísticos jerárquicos. La proximidad de categorías relacionadas (e.g., *ADP-DET* o *AUX-VERB*) en UMAP sugiere que el modelo captura gradientes sintácticos, no meras agrupaciones discretas. Estos resultados respaldan el uso de Spectral Clustering para análisis POS, priorizando su capacidad para modelar relaciones no lineales en espacios de alta dimensión, complementado con técnicas de reducción dimensional que preservan estructura local (*t-SNE*) o global-local (*UMAP*).
```
Comment (umoqnier) [inline practicas/fwgalde/P5/5_tech_evolution.py:1501]: Interesante, veo que rindió sus frutos buscar entre diversos métodos de clustering :100:

File: practicas/fwgalde/P5/5_tech_evolution.py [hunk 5]
```diff
@@ -0,0 +1,2170 @@
...
+- Reporta el accuracy y comparalo con los resultados de la RNN simple
+- Realiza un comentario sobre como impacta la arquitectura LSTM sobre el resultado obtenido
+
+### ***RNN LSTM***
+"""
+
+class LstmModel(nn.Module):
+    """Modelo LSTM para clasificación de secuencias con embeddings pre-entrenados.
+
+    :param embedding_table: Tensor de embeddings pre-entrenados
```
Comment (umoqnier) [inline practicas/fwgalde/P5/5_tech_evolution.py:1515]: Procura no mezclar estilos de docstrings, este es del estilo `reST` no el que usamos habitualmente en clase (`numpy`)

File: practicas/fwgalde/P5/5_tech_evolution.py [hunk 6]
```diff
@@ -0,0 +1,2170 @@
...
+                 bidirectional=False,
+                 dropout_rate=0.3):
+        super().__init__()
+        embedding_dim = embedding_table.size(dim=-1)
+        self.embedding = nn.Embedding.from_pretrained(
+            embedding_table,
+            freeze=freeze_embeddings,
+            padding_idx=0
+            )
+        self.lstm = nn.LSTM( # Cambio clave: nn.RNN -> nn.LSTM
```
Comment (umoqnier) [inline practicas/fwgalde/P5/5_tech_evolution.py:1554]: Nice!!!

File: practicas/fwgalde/P5/5_tech_evolution.py [hunk 7]
```diff
@@ -0,0 +1,2170 @@
...
+    torch.save({
+        'epoch': epoch,
+        'model_state_dict': lstm_model.state_dict(),
+        'optimizer_state_dict': optimizer.state_dict(),
+        'history': lstm_history
+    }, "pos_tagger_lstm_last_checkpoint.pth")
+
+# Guardar historial final
+torch.save(best_lstm_dict, path_lstm)
+torch.save(lstm_history, "pos_tagger_lstm.history")
```
Comment (umoqnier) [inline practicas/fwgalde/P5/5_tech_evolution.py:1702]: Hubiera estado bien que agregaras un link público con tu modelo para no tener que entrenar de mi lado jeje

File: practicas/fwgalde/P5/5_tech_evolution.py [hunk 8]
```diff
@@ -0,0 +1,2170 @@
...
+
+El incremento del 0.54% frente a la RNN estándar confirma que los mecanismos de *gates* (olvido, entrada y salida) de la LSTM mitigan eficazmente el problema de desvanecimiento de gradientes, permitiendo aprender patrones sintácticos complejos en secuencias largas. Comparado con modelos no secuenciales (FF, embeddings estáticos), la ganancia del 2.48-2.73% refleja la importancia de incorporar información temporal, irreductible a meras características léxicas o posicionales.  
+
+Aunque el costo computacional de la LSTM es mayor, su precisión justifica su uso en aplicaciones donde errores de tagging tienen consecuencias críticas (e.g., traducción automática o análisis sintáctico). Sin embargo, el modesto margen frente a la RNN (0.54%) sugiere que, para secuencias cortas o tareas menos dependientes de contexto histórico, arquitecturas más simples podrían ofrecer un equilibrio eficiencia-rendimiento aceptable. En síntesis, la LSTM emerge como el modelo óptimo para POS tagging en escenarios que exigen máxima precisión, gracias a su equilibrio entre capacidad de modelado secuencial y estabilidad en el entrenamiento.
```
Comment (umoqnier) [inline practicas/fwgalde/P5/5_tech_evolution.py:1735]: :+1:

File: practicas/fwgalde/P5/5_tech_evolution.py [hunk 9]
```diff
@@ -0,0 +1,2170 @@
...
+legend_centroid = Line2D([0],[0], marker='X', color='w', markerfacecolor=centroid_color,
+                        markersize=10, label='Centroides POS')
+
+plt.legend(handles=legend_words + [legend_centroid],
+           title="POS real", bbox_to_anchor=(1.05,1), loc='upper left')
+
+
+plt.show()
+
+"""Al igual que con t-SNE aquí resulta evidente como los valores de `rnn_model.linear.weight` se encuentran en un espacio dimensional distinto al de los embeddings léxicos. De ahí que este acercamiento no sea correcto cuando aplicamos una transformación no lineal para representar los datos."""
```
Comment (umoqnier) [inline practicas/fwgalde/P5/5_tech_evolution.py:2170]: Razonable, veo que debería ser más claro en cuanto a que me refería con embeddings de las etiquetas POS. Mala mía :p

**Review by umoqnier — APPROVED** [review]
# Calificación: 9 + EXTRA

**Por entrega tardía sé bajo un punto**

- Veo que profundizaste muchísimo en el clustering 📚
- Genial las anotaciones sobre que es lo que estás pensando y como interpretas los resultados
- En general resolviste bien la práctica e interpretaste correctamente tus resultados
